working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+# Horizontal inference
+
+Author: paulberry@google.com, Version: 1.0 (See [Changelog](#Changelog) at end)
+
+Horizontal inference allows certain arguments of an invocation (specifically
+function literals) to be type inferred in a context derived from the type of
+other arguments.
+
+## Motivation
+
+The key motivating example is the Iterable.fold method, which has the following
+signature:
+
+```dart
+abstract class Iterable<E> {
+  T fold<T>(T initialValue, T combine(T previousValue, E element)) { ... }
+}
+```
+
+A typical usage of `fold` today looks like this:
+
+```dart
+Iterable<int> values = ...;
+int largestValue = values.fold(0, (a, b) => a < b ? b : a);
+```
+
+Note the use of the explicit type `int` for `largestValue`.  We would like for
+the user to be able to drop this explicit type, and instead write:
+
+```dart
+Iterable<int> values = ...;
+var largestValue = values.fold(0, (a, b) => a < b ? b : a);
+```
+
+Today this doesn't work, because without the leading `int`, the downwards
+inference phase of type inference has no information with which to choose a
+preliminary type for the type parameter `T`, so the type context used for
+inference of the function literal `(a, b) => a < b ? b : a` is `? Function(?,
+int)`.  Hence, `a` gets assigned a static type of `Object?`, and this leads to a
+compile error at `a < b`.
+
+With horizontal inference, type inference first visits the `0`, and uses its
+static type to assign a preliminary type of `int` to the type parameter `T`.
+Then, when it proceeds to perform type inference on the function literal, the
+type context is `int Function(int, int)`.  This results in `a` being assigned a
+static type of `int`, which is what the user intends.
+
+## Type inference algorithm for invocations
+
+The full algorithm for type inference of invocations is expanded as follows.
+This algorithm is used for constructor invocations (including those in
+annotations, super-constructor calls, and constructor redirects), method
+invocations (including super calls), and invocations of function-typed
+expressions.
+
+1. Initial constraints: Create some constraints on type parameters by trying to
+   match the return type of the invocation target as a subtype of the incoming
+   type context.  (For a constructor invocation, the return type of the
+   invocation target is considered the raw uninstantiated type of the class
+   enclosing the constructor declaration.)
+
+2. Downwards inference: partially solve the set of type constraints accumulated
+   in step 1, to produce a preliminary mapping of type parameters to type
+   schemas.
+   
+3. Visit arguments (legacy): If experiment flag `inference-update-1` is not
```

Comment:
Agreed.  I've reworked the document mostly according to your suggestions.  Rather than include a `3-current` I've simply added a non-normative section near the end called "Differences from language version 2.18" that describes in prose how the current behavior differs.  I believe the difference is clear enough that it doesn't need to be spelled out formally.